### PR TITLE
[Pipelines] Add `MultiDiffusionText2PanoramaPipeline`

### PR DESCRIPTION
**Pull request by sayakpaul**
The main gotcha of this pipeline is that it doesn't (yet?) support PNDMScheduler likely because the generation loop has another loop for refining the view synthesis process. 

The main author (Omer) tested it with the DDIMScheduler and hence I kept it like so and added an error message. Happy to adjust the code to something better, feel free to suggest. 

Demo code:

```py
import torch
from diffusers import MultiDiffusionText2PanoramaPipeline, DDIMScheduler

model_ckpt = "stabilityai/stable-diffusion-2-base"
scheduler = DDIMScheduler.from_pretrained(model_ckpt, subfolder="scheduler")
pipe = MultiDiffusionText2PanoramaPipeline.from_pretrained(
    model_ckpt, scheduler=scheduler, torch_dtype=torch.float16
)

pipe = pipe.to("cuda")
prompt = "a photo of the dolomites"
image = pipe(prompt).images[0]
image
```

image

[Comparison Colab](https://colab.research.google.com/drive/1GFInXo3IWzS-VYh2sKSPq4wA5O7-a6uV?usp=sharing) (Omer's code and the pipeline results)

### CODE REVIEW

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+        Returns:
+            [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] or `tuple`:
+            [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] if `return_dict` is True, otherwise a `tuple.
+            When returning a tuple, the first element is a list with the generated images, and the second element is a
+            list of `bool`s denoting whether the corresponding generated image likely represents "not-safe-for-work"
+            (nsfw) content, according to the `safety_checker`.
+        """
+        # # 0. Default height and width to unet
+        # height = height or self.unet.config.sample_size * self.vae_scale_factor
+        # width = width or self.unet.config.sample_size * self.vae_scale_factor
```
Comment (sayakpaul) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:576]: To remove.

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 2]
```diff
@@ -0,0 +1,694 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:1]: We should also mention the original authors here.

File: src/diffusers/pipelines/__init__.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
     from .dit import DiTPipeline
     from .latent_diffusion import LDMSuperResolutionPipeline
     from .latent_diffusion_uncond import LDMPipeline
+    from .multidiffusion_panorama import MultiDiffusionText2PanoramaPipeline
```
Comment (patil-suraj) [inline src/diffusers/pipelines/__init__.py:24]: I think the pipeline can go in `stable_diffusion` directory. wdyt @patrickvonplaten

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+        scheduler: DDIMScheduler,
+        safety_checker: StableDiffusionSafetyChecker,
+        feature_extractor: CLIPFeatureExtractor,
+        requires_safety_checker: bool = True,
+    ):
+        super().__init__()
+
+        if isinstance(scheduler, PNDMScheduler):
+            logger.error("PNDMScheduler for this pipeline is currently not supported.")
+
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:99]: It would be cool to see how to support other schedulers. 

Also, if for now it only works with `DDIMScheduler`, we should raise an error if it's not an instance of `DDIMScheduler`, instead of just checking for `PNDMScheduler`

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+            tokenizer=tokenizer,
+            unet=unet,
+            scheduler=scheduler,
+            safety_checker=safety_checker,
+            feature_extractor=feature_extractor,
+        )
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.register_to_config(requires_safety_checker=requires_safety_checker)
+
+    def enable_vae_slicing(self):
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:176]: we should add `# Copied from ...` comments here since it's exactly similar to the methods in stable diffusion.

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+        for module in self.unet.modules():
+            if (
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
+
+    def _encode_prompt(
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:229]: same comment as above.

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:402]: Same comment as above.

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+
+        if prompt_embeds is not None and negative_prompt_embeds is not None:
+            if prompt_embeds.shape != negative_prompt_embeds.shape:
+                raise ValueError(
+                    "`prompt_embeds` and `negative_prompt_embeds` must have the same shape when passed directly, but"
+                    f" got: `prompt_embeds` {prompt_embeds.shape} != `negative_prompt_embeds`"
+                    f" {negative_prompt_embeds.shape}."
+                )
+
+    def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:449]: same comment as above.

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+            latents = latents.to(device)
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    # Credits to Omer Bar Tal for this method.
+    def get_views(self, panorama_height, panorama_width, window_size=64, stride=8):
+        panorama_height /= 8
+        panorama_width /= 8
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:469]: Could we add a comment/docstr here explaining what this function does?

File: src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py [hunk 10]
```diff
@@ -0,0 +1,694 @@
...
+        # views.
+        num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                count.zero_()
+                value.zero_()
+
+                # generate views
+                # credits to Omer Bar Tal for this utility.
+                for h_start, h_end, w_start, w_end in views:
```
Comment (patil-suraj) [inline src/diffusers/pipelines/multidiffusion_panorama/pipeline_multidiffusion_text2panorama.py:644]: It would be awesome to add a comment here to explain what's happening.

**Review by patil-suraj — COMMENTED** [review]
Awesome, thanks a lot for adding this pipeline so quickly! I left a few comments below, need to have a few more things before we can merge this:

- I think this pipeline can go in `stable_diffusion` directory. Let's see what @patrickvonplaten @pcuenca  thinks.
- We should add the `# Copied from ...` comment wherever possible.
- Should add a doc page for this with an example and abstract of the paper.
- And finally, we should add tests for this pipeline. 

Also would be cool to check if it works other schedulers , I think it does not work with PNDM because PNDM is a stateful scheduler and stores some information inside cf https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_pndm.py#L314, and this pipeline calls `step` multiple times for different views.

Happy to help with the schedulers and tests :)

**Comment by patrickvonplaten** [comment]
@patil-suraj why did we close this?

**Comment by sayakpaul** [comment]
> @patil-suraj why did we close this?

Replied internally via Slack.
